ambry-clustermap/src/main/java/com/github/ambry/clustermap/ClusterMapUtils.java

```diff
@@ -943,9 +946,47 @@ private boolean areAllLocalReplicasForPartitionUp(PartitionId partitionId) {
           return false;
         }
       }
+
+      // Generate a metric to track how often we choose a replica that is not globally available for writes.
+      if (!isPartitionEligibleForParanoidDurability(partitionId)) {
+        if (clusterManagerMetrics != null)
+          clusterManagerMetrics.paranoidDurabilityIneligibleReplicaCount.inc();
+      }
       return true;
     }
 
+    /**
+     * Checks whether there is at least one replica in every remote data center that is eligible for puts.
+     * @param partitionId the {@link PartitionId} to check.
+     * @return true if the given partition has at least one replica in every remote data center that is both up and
+     * in the right Helix state.
+     */
+    private boolean isPartitionEligibleForParanoidDurability(PartitionId partitionId) {
+      // Create a map where the keys are names of remote data centers and the values are all initialized to zero.
+      Map<String, Integer> validRemoteReplicasPerDc = partitionId.getReplicaIds().stream()
+          .filter(replica -> !replica.getDataNodeId().getDatacenterName().equals(this.localDatacenterName))
+          .collect(Collectors.groupingBy(replica -> replica.getDataNodeId().getDatacenterName(),
+              Collectors.reducing(0, replica -> 0, Integer::sum)));
+
+      if (validRemoteReplicasPerDc.isEmpty())
+        return false;
+
+      // For each remote data center, count the number of replicas that are not down and also in either LEADER or STANDBY state.
+      for (String dc : validRemoteReplicasPerDc.keySet()) {
+        Set<ReplicaId> eligibleReplicas = new HashSet<>();
```

Comment:
@litingulfs We discussed a way to make the multiple loops more efficient offline. If possible please update the code to make it cleaner.

Reply:
Thanks, the code has been cleaned up now.